tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,156 @@
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+    SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+    OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#pragma once
+
+namespace tiny_dnn {
+
+/* Class modelling a Tensor
+ */
+class Tensor {
+ public:
+    /* Default constructor for the Tensor class
+     * @param batch The size of the batch
+     * @param width The width of the tensor
+     * @param heigth The height of the tensor
+     * @param depth The number of channels
+     */
+    explicit Tensor(const cnn_size_t batch,
+                    const cnn_size_t width,
+                    const cnn_size_t height,
+                    const cnn_size_t depth) : shape_(4) {
+        init_data (batch, width, height, depth);
+        init_shape(batch, width, height, depth);
+    }
+    
+    // Returns the tensor shape
+    std::vector<cnn_size_t> shape() const { return shape_; }
+    
+    // Returns the value of a requested position in the tensor
+    float_t at(const cnn_size_t batch,  const cnn_size_t width,
+               const cnn_size_t height, const cnn_size_t depth) const {
+        return *access_data(batch, width, height, depth);
+    }
+
+    // Returns the pointer to a requested position in the tensor
+    float_t* ptr(const cnn_size_t batch,  const cnn_size_t width,
+                 const cnn_size_t height, const cnn_size_t depth) const {
+        return access_data(batch, width, height, depth);
+    }
+
+ private:
+    // Initializes the data buffer with zeroes
+    void init_data(const cnn_size_t batch,  const cnn_size_t width,
+                   const cnn_size_t height, const cnn_size_t depth) {
+        if (batch <= 0 || width <= 0 || height <= 0 || depth <= 0) {
+            nn_error("Tensor size must be a positive number");
+        }
+        data_cpu_ = std::make_shared<std::vector<float_t> >(
+            batch * width * height * depth, float_t(0));
+    }
+
+    // Initializes the shape vector
+    void init_shape(const cnn_size_t batch,  const cnn_size_t width,
+                    const cnn_size_t height, const cnn_size_t depth) {
+       shape_[0] = batch;  shape_[1] = width;
+       shape_[2] = height; shape_[3] = depth;
+    }
+
+    // Method to access to the tensor data.
+    // It checks if the requested position is feasible or not.
+    float_t* access_data(const cnn_size_t batch,
+                         const cnn_size_t width,
+                         const cnn_size_t height,
+                         const cnn_size_t depth) const {
+        if ((batch  < 0 || batch  > shape_[0]) ||
+            (width  < 0 || width  > shape_[1]) ||
+            (height < 0 || height > shape_[2]) ||
+            (depth  < 0 || depth  > shape_[3])) {
```

Comment:
@edgarriba 
size_t is always non-negative. We can remove first condition for each argument.